# 4.5: Data Consistency Checks

# 1: Importing Data

## 1.1: Imporing libraries

In [2]:
# Importing libraries for the analysis
import pandas as pd
import numpy as np
import os

## 1.2: Importing Data Sets

In [5]:
#Importing the necessary data sets for the task.
#Creating a path for the import of data.
path = r'/Users/daniel/Desktop/LEBEN/Data Analyst/Data Analytics Course/Data Immersion/Achievement 4/Master Folder - Instacart Basket Analysis 09.09.2024'

In [7]:
#Importing the products.csv data from the original data folder
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
#Importing the orders_wrangled.csv data from the prepared data folder
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 2: Data Consitency Check - df_prods

## 2.1: Missing values

### 2.1.1: Finding missing values within the df_prods

In [14]:
#Finding missing within the df_prods. The isnull() finds the missing values, and the sum() obtains the total number.
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
#Create a new data frame that shows the missing values of the df_prods.
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
# Checking for the number of rows of the set before cleaning to see if the function applies correctly
df_prods.shape

(49693, 5)

### 2.1.2: Addressing missing values

In [23]:
#Creating a new data frame that excludes the rows with the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

__The total rows are 49677. Before cleaning (49693 rows) - After cleaning (49677 rows) = 16 rows less. The function was applied correctly.__

## 2.2: Duplicate Values

### 2.2.1: Finding duplicate values within the df_prods

In [30]:
#Finding the duplicate values by using the duplicated() function within the df_prods.
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [32]:
# Show the duplicate values within the df_prods.
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### 2.1.2: Addressing duplicated values

In [35]:
#Addressing duplicates by creating a new data frame without the duplicated values.
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [37]:
df_prods_clean_no_dups.shape

(49672, 5)

__The total rows are 49672. Before cleaning (49677 rows) - After cleaning (49672 rows) = 5 rows less. The function was apllied correclty.__

# 3: Data Consitency Check - df_ords

## 3.1: Descriptive analysis and short interpretation

In [42]:
#conduct a descriptive analysis on the data frame df_ords and check if there is anything off about the data
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_last_order  
count            3.421083e+06       3.421083e+06           3.214874e+06  
mean             2.776219e+00       1.345202e+01           1.111484e+01  
std              2.046829e+00       4.226088e+00           9.206737e+00  
min              0.000000e+00       0.000000e+00           0.000000e+00  
25%              1.000000e+00       1.000000e+01           4.000000e+00  
50%              3.000000e+00       1.300000e+01           7.000000e+00  
75%              5.000000e+00       1.600000e+01           1.500000e+01  
max              6.000000e+00       2.300000e+01           3.000000e+01

__Interpretation of the df_ords: 
<br>
<br>
days_since_prior_order: This column has a lower count of values compared to others, suggesting some missing data. It includes a value of zero, which indicates that some orders could be placed on the same day. The maximum value is 30, which might suggest that orders are tracked for up to 30 days or one month.
<br>
<br>
order_number: The maximum value in this column is 100, which could imply that the maximum number of items per order is capped at 100.
<br>
<br>
order_day_of_week: This column has values ranging from 0 to 6, corresponding to the seven days of the week.
<br>
<br>
order_hour_of_day: This column has values ranging from 0 to 23, representing the 24 hours of the day.__

## 3.2: Checking for mixed_type data

In [46]:
#Checking the df_ords for mixed-type data, by using a for-loop function to check every column for mixed data types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].map(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

__Since there is not output, the df_ords contains just single data type columns.__

## 3.3: Missing values

### 3.3.1: Finding missing values within the df_ords

In [51]:
#Finding missing within the df_ords. The isnull() finds the missing values, and the sum() obtains the total number.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_last_order     206209
dtype: int64

In [57]:
#Creating a new data frame for the missing values here
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [59]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
11               11   2168274        2             1                       2   
26               26   1374495        3             1                       1   
39               39   3343014        4             1                       6   
45               45   2717275        5             1                       3   
...             ...       ...      ...           ...                     ...   
3420930     3420930    969311   206205             1                       4   
3420934     3420934   3189322   206206             1                       3   
3421002     3421002   2166133   206207             1                       6   
3421019     3421019   2227043   206208             1                       1   
3421069     3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

__Explanation of the missing values__: The 'days_since_last_order' column counts a total of 206209 missing values. One explanation could be that the customers who placed the order deleted their accounts afterward. so the system can not count any days since the last order, because a new order can not be placed due to the deletion of the account. Another possibility could be the NaN values represent new customers or/and the first order of new customers.

### 3.3.2: Addressing missing values within the df_ords

In [65]:
#Assuming that the NaN values represent new customers or the first order of new customers, it is appropriate to replace the NaN values with 0
df_ords['days_since_last_order'] = df_ords['days_since_last_order'].fillna(0)

In [69]:
# Recheck the df_ords again for missing values, and see if the fillna function is applied probably.
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_the_week    0
order_hour_of_day         0
days_since_last_order     0
dtype: int64

__Explanation of the method__: I decided to replace NaN with 0 because it represents the data entry of new customers or the first order of customers. Considering that, the values within the column will adjust after the second order is placed.

## 3.4: Duplicate Values

### 3.4.1: Finding duplicate values within the df_ords

In [74]:
#Checking for duplicate values within the df_ords.
df_ords_dups = df_ords[df_ords.duplicated()]

In [76]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

__Explanation__: No duplicates are found within the dataset. In case duplicate values would have been, I would drop them to create a new data frame without any duplicates

# 5: Export the Data

In [88]:
#Export the df_ords dataframe as "orders_cleaned.csv"
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [81]:
#Export the df_prods_clean_no_dups as "prods_cleaned.csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))